README.md
# Yandex Maps API Polygonmap Module

Yandex.Maps API module for data visualization.

**Polygonmap** is a graphical representation of some spatial data, where depending on the number of entered points polygons are painted in different colors.
`Polygonmap` class allows to construct and display such representations over geographical maps.

## Loading

1. Put module source code ([polygonmap.min.js](https://github.com/yandex-shri-fx-team/ymaps-polygonmap/blob/master/umd/polygonmap.min.js)) on your CDN.

2. Load both [Yandex Maps JS API 2.1](http://api.yandex.com/maps/doc/jsapi/) and module source code by adding following code into &lt;head&gt; section of your page:

   ```html
   <script src="http://api-maps.yandex.ru/2.1/?lang=ru_RU" type="text/javascript"></script>
   <!-- Change my.cdn.tld to your CDN host name -->
   <script src="http://my.cdn.tld/polygonmap.min.js" type="text/javascript"></script>
   ```

   If you use [GeoJSON](http://geojson.org) data:

   ```html
   <script src="http://api-maps.yandex.ru/2.1/?lang=ru_RU&coordOrder=longlat" type="text/javascript"></script>
   <!-- Change my.cdn.tld to your CDN host name -->
   <script src="http://my.cdn.tld/polygonmap.min.js" type="text/javascript"></script>
   ```

   If you use [npm](https://www.npmjs.com):

   ```html
   <script src="http://api-maps.yandex.ru/2.1/?lang=ru_RU" type="text/javascript"></script>
   ```

   ```bash
   npm i --save ymaps-polygonmap
   ```

   ```js
   require('ymaps-polygonmap');

   // Or with babel
   import 'ymaps-polygonmap';
   ```

3. Get access to module functions by using [ymaps.modules.require](http://api.yandex.ru/maps/doc/jsapi/2.1/ref/reference/modules.require.xml) method:

   ```js
   ymaps.modules.require(['Polygonmap'], function (Polygonmap) {
        var polygonmap = new Polygonmap();
   });
   ```

<a name="module_Polygonmap"></a>

## Polygonmap
Polygonmap module.

**Requires**: <code>module:option.Manager</code>, <code>module:ObjectManager</code>  

* [Polygonmap](#module_Polygonmap)
    * [Polygonmap](#exp_module_Polygonmap--Polygonmap) ⏏
        * [new Polygonmap([data], [options])](#new_module_Polygonmap--Polygonmap_new)
        * [.getData()](#module_Polygonmap--Polygonmap+getData) ⇒ <code>Object</code>
        * [.setData(data)](#module_Polygonmap--Polygonmap+setData) ⇒ <code>Polygonmap</code>
        * [.getMap()](#module_Polygonmap--Polygonmap+getMap) ⇒ <code>Map</code>
        * [.setMap(map)](#module_Polygonmap--Polygonmap+setMap) ⇒ <code>Polygonmap</code>
        * [.destroy()](#module_Polygonmap--Polygonmap+destroy)

<a name="exp_module_Polygonmap--Polygonmap"></a>

### Polygonmap ⏏
**Kind**: Exported class  
<a name="new_module_Polygonmap--Polygonmap_new"></a>

#### new Polygonmap([data], [options])

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [data] | <code>Object</code> |  | Polygons and points. |
| data.polygons | <code>Object</code> |  | GeoJSON FeatureCollections. |
| data.points | <code>Object</code> |  | GeoJSON FeatureCollections. |
| [options] | <code>Object</code> |  | Options for customization. |
| [options.mapper] | <code>function</code> | <code>defaultMapper</code> | Function of iterative transformation of features. |
| [options.fillBy] | <code>string</code> | <code>&quot;points&quot;</code> | Calculate the color by points | weight. |
| [options.fillByWeightProp] | <code>string</code> | <code>&quot;weight&quot;</code> | Prop name in data object, for weight value. If fillBy is "weight". |
| [options.fillByWeightType] | <code>string</code> | <code>&quot;middle&quot;</code> | Type of calculate color by weight. Can be middle | maximum |
| [options.colorRanges] | <code>number</code> \| <code>array</code> | <code>4</code> | Count of ranges or array of custom ranges. |
| [options.colorScheme] | <code>string</code> \| <code>array</code> | <code>&quot;[&#x27;#e66a54&#x27;, &#x27;#ce4356&#x27;, &#x27;#ab2960&#x27;, &#x27;#571756&#x27;]&quot;</code> | Preset for colorize or array of custom colors. |
| [options.fillOpacity] | <code>number</code> | <code>0.8</code> | Opacity of polygon. |
| [options.fillColorEmptyPolygon] | <code>string</code> | <code>&quot;#f4f6f8&quot;</code> | Color of polygon where points count equal 0. |
| [options.strokeColor] | <code>string</code> | <code>&quot;#fff&quot;</code> | Color of polygon stroke. |
| [options.strokeWidth] | <code>number</code> | <code>1</code> | Width of polygon stroke. |
| [options.showLegend] | <code>boolean</code> | <code>true</code> | Flag to show color legend. |
| [options.legendTemplate] | <code>function</code> | <code>colorLegend.defaultTemplate</code> | Receives object {color: value} returns html legend template. |
| [options.legendPosition] | <code>object</code> | <code>top: 10, right: 10</code> | Position of legend, you can only change the top or bottom and right or left. |
| [options.filter] | <code>function</code> |  | Function for custom filter polygons with points. |
| [options.filterEmptyPolygons] | <code>boolean</code> | <code>false</code> | Flag for show polygon with count of points equal 0. |
| [options.onMouseEnter] | <code>function</code> | <code>defaultOnMouseEnter</code> | Handler for mouseEnter event. |
| [options.onMouseLeave] | <code>function</code> | <code>defaultOnMouseLeave</code> | Handler for mouseLeave event. |
| [options.onClick] | <code>function</code> | <code>defaultOnClick</code> | Handler for click event. |
| [options.balloonContent] | <code>function</code> | <code>defaultBalloonContent</code> | Function for render content of baloon. Recieves object with properties of polygon. |
| [options.fillColorHover] | <code>string</code> |  | Color of polygon on polygon hover. |
| [options.fillOpacityHover] | <code>number</code> | <code>0.9</code> | Number of opacity on polygon hover. |
| [options.strokeColorHover] | <code>string</code> |  | Color of polygon stroke on polygon hover. |
| [options.strokeWidthHover] | <code>number</code> | <code>2</code> | Number of stroke width on polygon hover. |
| [options.fillColorActive] | <code>string</code> |  | Color of polygon on polygon active. |
| [options.fillOpacityActive] | <code>number</code> | <code>0.9</code> | Number of opacity on polygon active. |
| [options.strokeColorActive] | <code>string</code> |  | Color of polygon stroke on polygon active. |
| [options.strokeWidthActive] | <code>number</code> | <code>2</code> | Number of stroke width on polygon active. |
| [options.interactivity] | <code>boolean</code> | <code>true</code> | Flag for enable interactivity. |

<a name="module_Polygonmap--Polygonmap+getData"></a>

#### polygonmap.getData() ⇒ <code>Object</code>
Get the data, polygons and points.

**Kind**: instance method of [<code>Polygonmap</code>](#exp_module_Polygonmap--Polygonmap)  
**Returns**: <code>Object</code> - Polygons and points.  
**Access**: public  
<a name="module_Polygonmap--Polygonmap+setData"></a>

#### polygonmap.setData(data) ⇒ <code>Polygonmap</code>
Set the data, polygons and points.

**Kind**: instance method of [<code>Polygonmap</code>](#exp_module_Polygonmap--Polygonmap)  
**Returns**: <code>Polygonmap</code> - Self-reference.  
**Access**: public  

| Param | Type | Description |
| --- | --- | --- |
| data | <code>Object</code> | Polygons and points. |
| data.polygons | <code>Object</code> | GeoJSON FeatureCollections. |
| data.points | <code>Object</code> | GeoJSON FeatureCollections. |

<a name="module_Polygonmap--Polygonmap+getMap"></a>

#### polygonmap.getMap() ⇒ <code>Map</code>
Get the Map instance.

**Kind**: instance method of [<code>Polygonmap</code>](#exp_module_Polygonmap--Polygonmap)  
**Returns**: <code>Map</code> - Reference to Map instance.  
**Access**: public  
<a name="module_Polygonmap--Polygonmap+setMap"></a>

#### polygonmap.setMap(map) ⇒ <code>Polygonmap</code>
Set Map instance to render Polygonmap object.

**Kind**: instance method of [<code>Polygonmap</code>](#exp_module_Polygonmap--Polygonmap)  
**Returns**: <code>Polygonmap</code> - Self-reference.  
**Access**: public  

| Param | Type | Description |
| --- | --- | --- |
| map | <code>Map</code> | Map instance. |

<a name="module_Polygonmap--Polygonmap+destroy"></a>

#### polygonmap.destroy()
Destructs Polygonmap instance.

**Kind**: instance method of [<code>Polygonmap</code>](#exp_module_Polygonmap--Polygonmap)  
**Access**: public  

## Examples

### Displaying polygonmap over geographical map

```js
ymaps.modules.require(['Polygonmap'], function (Polygonmap) {
    const dataPolygons = {
            type: 'FeatureCollection',
            features: [{
                id: 'id3',
                type: 'Feature',
                geometry: {
                    type: 'Polygon',
                    coordinates: [
                        [37.782051, -122.445068]
                        [37.782051, -122.445568]
                        [37.782951, -122.445068]
                        [37.782951, -122.445568]
                    ]
                }
            }]
        };
    const dataPoints = {
            type: 'FeatureCollection',
            features: [{
                id: 'id1',
                type: 'Feature',
                geometry: {
                    type: 'Point',
                    coordinates: [37.782551, -122.445368]
                }
            }, {
                id: 'id2',
                type: 'Feature',
                geometry: {
                    type: 'Point',
                    coordinates: [37.782745, -122.444586]
                }
            }]
        };
    const data = {polygons: dataPolygons, points: dataPoints};
    const polygonmap = new Polygonmap(data);

    polygonmap.setMap(myMap);
});
```

### Updating polygonmap data

```js
ymaps.modules.require(['Polygonmap'], function (Polygonmap) {
    const data = {polygons: dataPolygons, points: dataPoints};
    const polygonmap = new Polygonmap(data);

    polygonmap.setMap(myMap);

    const newData = {polygons: newDataPolygons, points: newDataPoints};

    polygonmap.setData(newData);
});
```

### Changing polygonmap representation options

```js
ymaps.modules.require(['Polygonmap'], function (Polygonmap) {
    const data = {polygons: dataPolygons, points: dataPoints};
    const polygonmap = new Polygonmap(data);

    polygonmap.options.set('mapper', (feature) => {
        const {pointsCount, pointsCountMaximum} = feature.properties;
        const k = pointsCount / pointsCountMaximum;

        feature.options = {fillColor: `rgba(0, 0, 0, ${k})`};

        return feature;
    });
    polygonmap.setMap(myMap);
});
```

## Demo

- https://yandex-shri-fx-team.github.io/ymaps-polygonmap


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.1.0"></a>
# [1.1.0](https://github.com/yandex-shri-fx-team/ymaps-polygonmap/compare/v1.0.1...v1.1.0) (2018-07-08)


### Bug Fixes

* **colorize:** reverse colorScheme, set from min to max order ([c7f3c07](https://github.com/yandex-shri-fx-team/ymaps-polygonmap/commit/c7f3c07))


### Features

* add auto calculate colorRanges/colorScheme ([734c7b0](https://github.com/yandex-shri-fx-team/ymaps-polygonmap/commit/734c7b0))
* **module:** add fillColor, strokeColor options for hover, active ([12d807d](https://github.com/yandex-shri-fx-team/ymaps-polygonmap/commit/12d807d))
* **module:** add fillColor, strokeColor options for hover, active ([1fc7648](https://github.com/yandex-shri-fx-team/ymaps-polygonmap/commit/1fc7648))
* **module:** add support for colorRangesMinimum ([cb7bffa](https://github.com/yandex-shri-fx-team/ymaps-polygonmap/commit/cb7bffa))



<a name="1.0.1"></a>
## [1.0.1](https://github.com/yandex-shri-fx-team/ymaps-polygonmap/compare/v1.0.0...v1.0.1) (2018-06-11)


### Bug Fixes

* **module:** edit default options ([d813d2e](https://github.com/yandex-shri-fx-team/ymaps-polygonmap/commit/d813d2e))
* **module:** fix jsdoc default values and optional syntax ([6b52914](https://github.com/yandex-shri-fx-team/ymaps-polygonmap/commit/6b52914))
* **module:** update jdoc default values ([9f67d86](https://github.com/yandex-shri-fx-team/ymaps-polygonmap/commit/9f67d86))


src/Polygonmap.js
import normalizeFeature from './utils/normalizeFeature';
import defaultMapper from './utils/defaultMapper';
import defaultFilter from './utils/defaultFilter';
import defaultOnMouseEnter from './utils/defaultOnMouseEnter';
import defaultOnMouseLeave from './utils/defaultOnMouseLeave';
import defaultBalloonContent from './utils/defaultBalloonContent';
import defaultOnClick from './utils/defaultOnClick';
import inside from './utils/inside';
import Colorize from './utils/colorize';
import colorLegend from './utils/colorLegend';

/**
 * Polygonmap module.
 *
 * @module Polygonmap
 * @requires option.Manager
 * @requires ObjectManager
 */
ymaps.modules.define('Polygonmap', [
    'meta',
    'option.Manager',
    'ObjectManager'
], (provide, meta, OptionManager, ObjectManager) => {
    /**
     * @param {Object} [data] Polygons and points.
     * @param {Object} data.polygons GeoJSON FeatureCollections.
     * @param {Object} data.points GeoJSON FeatureCollections.
     * @param {Object} [options] Options for customization.
     * @param {function} [options.mapper=defaultMapper] Function of iterative transformation of features.
     * @param {string} [options.fillBy=points] Calculate the color by points | weight.
     * @param {string} [options.fillByWeightProp=weight] Prop name in data object, for weight value.
     * If fillBy is "weight".
     * @param {string} [options.fillByWeightType=middle] Type of calculate color by weight. Can be middle | maximum
     * @param {number|array} [options.colorRanges=4] Count of ranges or array of custom ranges.
     * @param {string|array} [options.colorScheme=['#e66a54', '#ce4356', '#ab2960', '#571756']]
     * Preset for colorize or array of custom colors.
     * @param {number} [options.fillOpacity=0.8] Opacity of polygon.
     * @param {string} [options.fillColorEmptyPolygon=#f4f6f8] Color of polygon
     * where points count equal 0.
     * @param {string} [options.strokeColor=#fff] Color of polygon stroke.
     * @param {number} [options.strokeWidth=1] Width of polygon stroke.
     * @param {boolean} [options.showLegend=true] Flag to show color legend.
     * @param {function} [options.legendTemplate=colorLegend.defaultTemplate] Receives object {color: value}
     * returns html legend template.
     * @param {object} [options.legendPosition=top: 10, right: 10] Position of legend,
     * you can only change the top or bottom and right or left.
     * @param {function} [options.filter=undefined] Function for custom filter polygons with points.
     * @param {boolean} [options.filterEmptyPolygons=false] Flag for show polygon with count of points equal 0.
     * @param {function} [options.onMouseEnter=defaultOnMouseEnter] Handler for mouseEnter event.
     * @param {function} [options.onMouseLeave=defaultOnMouseLeave] Handler for mouseLeave event.
     * @param {function} [options.onClick=defaultOnClick] Handler for click event.
     * @param {function} [options.balloonContent=defaultBalloonContent] Function for render content of baloon.
     * Recieves object with properties of polygon.
     * @param {string} [options.fillColorHover] Color of polygon on polygon hover.
     * @param {number} [options.fillOpacityHover=0.9] Number of opacity on polygon hover.
     * @param {string} [options.strokeColorHover] Color of polygon stroke on polygon hover.
     * @param {number} [options.strokeWidthHover=2] Number of stroke width on polygon hover.
     * @param {string} [options.fillColorActive] Color of polygon on polygon active.
     * @param {number} [options.fillOpacityActive=0.9] Number of opacity on polygon active.
     * @param {string} [options.strokeColorActive] Color of polygon stroke on polygon active.
     * @param {number} [options.strokeWidthActive=2] Number of stroke width on polygon active.
     * @param {boolean} [options.interactivity=true] Flag for enable interactivity.
     *
     * @alias module:Polygonmap
     */
    class Polygonmap {
        constructor(data, options = {}) {
            const defaultOptions = new OptionManager({
                mapper: defaultMapper,
                fillBy: 'points',
                fillByWeightProp: 'weight',
                fillByWeightType: 'middle',
                colorRangesMinimum: 0,
                colorRanges: 4,
                colorScheme: [
                    '#e66a54',
                    '#ce4356',
                    '#ab2960',
                    '#571756'
                ],
                fillOpacity: 0.8,
                fillColorEmptyPolygon: '#f4f6f8',
                strokeColor: '#fff',
                strokeWidth: 1,
                showLegend: true,
                legendTemplate: colorLegend.defaultTemplate,
                legendPosition: {
                    top: 10,
                    right: 10
                },
                // Since the default filter for empty polygons is disabled by default,
                // this option will be undefined.
                filter: undefined,
                filterEmptyPolygons: false,
                onMouseEnter: defaultOnMouseEnter,
                onMouseLeave: defaultOnMouseLeave,
                onClick: defaultOnClick,
                balloonContent: defaultBalloonContent,
                fillColorHover: undefined,
                fillOpacityHover: 0.9,
                strokeColorHover: '#fff',
                strokeWidthHover: 2,
                fillColorActive: undefined,
                fillOpacityActive: 0.9,
                strokeColorActive: '#fff',
                strokeWidthActive: 2,
                interactivity: true
            });

            this._initOptions(options, defaultOptions);
            this.setData(data);
        }

        /**
         * Get the data, polygons and points.
         *
         * @public
         * @returns {Object} Polygons and points.
         */
        getData() {
            return this._data || null;
        }

        /**
         * Set the data, polygons and points.
         *
         * @public
         * @param {Object} data Polygons and points.
         * @param {Object} data.polygons GeoJSON FeatureCollections.
         * @param {Object} data.points GeoJSON FeatureCollections.
         * @returns {Polygonmap} Self-reference.
         */
        setData(data) {
            this._data = data;

            if (data) {
                this._data = {
                    points: {type: 'FeatureCollection', features: []},
                    polygons: {type: 'FeatureCollection', features: []}
                };
                this._prepare(data);
                this._initObjectManager();
            }

            return this;
        }

        /**
         * Get the Map instance.
         *
         * @public
         * @returns {Map} Reference to Map instance.
         */
        getMap() {
            return this._map;
        }

        /**
         * Set Map instance to render Polygonmap object.
         *
         * @public
         * @param {Map} map Map instance.
         * @returns {Polygonmap} Self-reference.
         */
        setMap(map) {
            if (this._map !== map) {
                this._map = map;

                if (map && this._data) {
                    this._render();
                }
            }

            return this;
        }

        /**
         * Destructs Polygonmap instance.
         *
         * @public
         */
        destroy() {
            this.setData(null);
            this.objectManager.removeAll();
            this._legendControl.setParent(null);
            this.balloon.close();
            this.setMap(null);
        }

        /**
         * Prepare for render Polygonmap.
         *
         * @private
         * @param {Object} data Polygons and points.
         * @param {Object} data.polygons GeoJSON FeatureCollections.
         * @param {Object} data.points GeoJSON FeatureCollections.
         * @returns {undefined}
         */
        _prepare(data) {
            if (
                data.polygons.type === 'FeatureCollection' &&
                data.points.type === 'FeatureCollection'
            ) {
                const polygonFeatures = data.polygons.features;

                const fillBy = this.options.get('fillBy');
                const fillByWeightType = this.options.get('fillByWeightType');
                const fillByWeightProp = this.options.get('fillByWeightProp');

                let pointFeatures = data.points.features;

                for (let i = 0; i < polygonFeatures.length; i++) {
                    const restPointFeatures = [];
                    const polygonFeature = normalizeFeature(polygonFeatures[i], meta, {id: i});

                    let pointsCount = 0;
                    let pointsWeight = 0;

                    for (let j = 0; j < pointFeatures.length; j++) {
                        let pointFeature;

                        if (i === 0) {
                            pointFeature = normalizeFeature(pointFeatures[j], meta, {id: j});
                            this._data.points.features.push(pointFeature);
                        } else {
                            pointFeature = pointFeatures[j];
                        }

                        if (inside(polygonFeature.geometry, pointFeature.geometry)) {
                            pointsCount++;
                            pointsWeight += pointFeature.properties[fillByWeightProp];
                        } else {
                            restPointFeatures.push(pointFeature);
                        }
                    }

                    pointFeatures = restPointFeatures;

                    if (fillBy === 'weight') {
                        pointsWeight = fillByWeightType === 'middle' ?
                            pointsWeight === 0 && pointsCount === 0 ?
                                0 :
                                pointsWeight / pointsCount :
                            pointsWeight;
                    }

                    polygonFeature.properties = polygonFeature.properties || {};
                    polygonFeature.properties.pointsCount = pointsCount;
                    polygonFeature.properties.pointsWeight = pointsWeight;

                    this._calculator(polygonFeature, {fillBy});

                    this._data.polygons.features.push(polygonFeature);
                }
            }
        }

        /**
         * Calculator for feture values.
         *
         * @param {Object} feature GeoJSON FeatureCollection.
         * @param {Object} options Options for calculating.
         * @private
         */
        _calculator(feature, options) {
            const {pointsCount} = feature.properties;
            const {pointsWeight} = feature.properties;
            const {fillBy} = options;

            if (
                typeof this.pointsCountMinimum === 'undefined' ||
                pointsCount <= this.pointsCountMinimum
            ) {
                this.pointsCountMinimum = pointsCount;
            }

            if (
                typeof this.pointsCountMaximum === 'undefined' ||
                pointsCount >= this.pointsCountMaximum
            ) {
                this.pointsCountMaximum = pointsCount;
            }

            if (fillBy === 'weight') {
                if (
                    typeof this.pointsWeightMinimum === 'undefined' ||
                    pointsWeight <= this.pointsWeightMinimum
                ) {
                    this.pointsWeightMinimum = pointsWeight;
                }

                if (
                    typeof this.pointsWeightMaximum === 'undefined' ||
                    pointsWeight >= this.pointsWeightMaximum
                ) {
                    this.pointsWeightMaximum = pointsWeight;
                }
            }
        }

        /**
         * Render ObjectManager.
         *
         * @private
         */
        _render() {
            if (this.options.get('interactivity')) {
                this._initInteractivity();
            }

            this._map.geoObjects.add(this.objectManager);
            colorLegend.init(this);
        }

        /**
         * Init Options.
         *
         * @param {Object} options Options.
         * @param {Object} defaultOptions Default options.
         * @private
         */
        _initOptions(options, defaultOptions) {
            this.options = new OptionManager(options, defaultOptions);

            const mapper = this.options.get('mapper');
            const filterEmptyPolygons = this.options.get('filterEmptyPolygons');
            const colorScheme = this.options.get('colorScheme');
            const colorRanges = this.options.get('colorRanges');
            const onMouseEnter = this.options.get('onMouseEnter');
            const onMouseLeave = this.options.get('onMouseLeave');
            const onClick = this.options.get('onClick');

            this.options.set('mapper', mapper.bind(this));

            if (filterEmptyPolygons) {
                this.options.set('filter', defaultFilter.bind(this));
            }

            if (
                typeof options.colorRanges === 'undefined' &&
                colorRanges !== colorScheme.length
            ) {
                this.options.set('colorRanges', colorScheme.length);
            }

            this.options.set('onMouseEnter', onMouseEnter.bind(this));
            this.options.set('onMouseLeave', onMouseLeave.bind(this));
            this.options.set('onClick', onClick.bind(this));
        }

        /**
         * Init ObjectManager.
         *
         * @private
         */
        _initObjectManager() {
            const mapper = this.options.get('mapper');
            const filter = this.options.get('filter');
            const fillBy = this.options.get('fillBy');
            const fillByWeight = fillBy === 'weight';
            const colorRangesMinimum = this.options.get('colorRangesMinimum');

            if (colorRangesMinimum === 'min') {
                delete this.pointsCountMinimum;
                delete this.pointsWeightMinimum;

                const reducer = (acc, feature) => {
                    if (!filter || filter(feature)) {
                        this._calculator(feature, {fillBy});

                        acc.push(feature);
                    }

                    return acc;
                };

                this._data.polygons.features = this._data.polygons.features.reduce(reducer, []);
            }

            try {
                this.colorize = new Colorize(
                    colorRangesMinimum === 'min' ?
                        fillByWeight ? this.pointsWeightMinimum : this.pointsCountMinimum :
                        colorRangesMinimum,
                    fillByWeight ? this.pointsWeightMaximum : this.pointsCountMaximum,
                    {
                        colorScheme: this.options.get('colorScheme'),
                        colorRanges: this.options.get('colorRanges')
                    }
                );
            } catch (error) {
                console.error(error);
            }

            const reducer = (acc, feature) => {
                if (colorRangesMinimum !== 'min' || !filter || filter(feature)) {
                    if (mapper) {
                        acc.push(mapper(feature));
                    } else {
                        acc.push(feature);
                    }

                    this._calculator(feature, {fillBy});
                }

                return acc;
            };

            this._data.polygons.features = this._data.polygons.features.reduce(reducer, []);

            this.objectManager = new ObjectManager();
            this.objectManager.add(this._data.polygons);
        }

        /**
         * Init interactivity.
         *
         * @private
         */
        _initInteractivity() {
            this._prevObjectId = null;
            this.balloon = new ymaps.Balloon(this._map);

            const onMouseEnter = this.options.get('onMouseEnter');
            const onMouseLeave = this.options.get('onMouseLeave');
            const onClick = this.options.get('onClick');

            this.objectManager.events.add('mouseenter', onMouseEnter);

            this.objectManager.events.add('mouseleave', onMouseLeave);

            this.balloon.options.setParent(this._map.options);

            this.objectManager.events.add('click', onClick);
        }
    }

    provide(Polygonmap);
});
